pkg/resourcemanager/controller/csrapprover/reconciler.go

```diff
@@ -166,3 +224,43 @@ func (r *Reconciler) mustApprove(ctx context.Context, csr *certificatesv1.Certif
 
 	return "all checks passed", true, nil
 }
+
+func (r *Reconciler) mustApproveKubeAPIServerClient(ctx context.Context, csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) (string, decision, error) {
+	if !strings.HasPrefix(x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix) {
+		return fmt.Sprintf("commonName %q is not prefixed with %q", x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix), csrNoOpinion, nil
+	}
+
+	machineName := strings.TrimPrefix(x509cr.Subject.CommonName, v1beta1constants.NodeAgentUserNamePrefix)
+	machine := &machinev1alpha1.Machine{}
+	if err := r.SourceClient.Get(ctx, client.ObjectKey{Namespace: r.Config.MachineNamespace, Name: machineName}, machine); err != nil {
+		if apierrors.IsNotFound(err) {
+			return fmt.Sprintf("machine %q does not exist", machineName), csrDenied, nil
+		}
+		return "", csrNoOpinion, err
+	}
+
+	switch {
+	case strings.HasPrefix(csr.Spec.Username, "system:bootstrap:"):
+		if nodeName, found := machine.Labels[machinev1alpha1.NodeLabelKey]; found {
+			node := &corev1.Node{}
+			if err := r.TargetClient.Get(ctx, client.ObjectKey{Name: nodeName}, node); err == nil {
+				return fmt.Sprintf("Cannot use bootstrap token. Gardener-node-agent for machine %q is already bootstrapped", machineName), csrDenied, nil
```

Comment:
```suggestion
			if err := r.TargetClient.Get(ctx, client.ObjectKey{Name: nodeName}, &corev1.Node{}); err == nil {
```suggestion
				return fmt.Sprintf("Cannot use bootstrap token since gardener-node-agent for machine %q is already bootstrapped", machineName), csrDenied, nil
```
```